source/mesh_deformation/diffusion.cc

```diff
@@ -0,0 +1,591 @@
+
+                // Only apply diffusion to the requested boundaries
+                if (boundary_ids.find(boundary_indicator) == boundary_ids.end())
+                  continue;
+
+                // Recompute values, gradients, etc on the faces
+                fs_fe_face_values.reinit (fscell, face_no);
+
+                // Get the global numbers of the local DoFs of the mesh deformation cell
+                fscell->get_dof_indices (cell_dof_indices);
+
+                // Extract the displacement values
+                fs_fe_face_values[extract_vertical_displacements].get_function_values (displacements, displacement_values);
+
+                // Extract the initial topography values
+                fs_fe_face_values[extract_initial_topography].get_function_values (initial_topography, initial_topography_values);
+
+                // Reset local rhs and matrix
+                cell_vector = 0;
+                cell_matrix = 0;
+
+                // Loop over the quadrature points of the current face
+                for (unsigned int point=0; point<n_fs_face_q_points; ++point)
+                  {
+                    // Get the gravity vector to compute the outward direction of displacement
+                    Tensor<1,dim> direction = -(this->get_gravity_model().gravity_vector(fs_fe_face_values.quadrature_point(point)));
+                    // Normalize direction vector
+                    if (direction.norm() > 0.0)
+                      direction *= 1./direction.norm();
+                    // TODO this is only correct for box geometries
+                    else
+                      direction[dim-1] = 1.;
+
+                    // Compute the total displacement in the gravity direction,
+                    // i.e. the initial topography + any additional mesh displacement.
+                    const double displacement = direction * (displacement_values[point] + initial_topography_values[point]);
+
+                    // To project onto the tangent space of the surface,
+                    // we define the projection P:= I- n x n,
+                    // with I the unit tensor and n the unit normal to the surface.
+                    // The surface gradient then is P times the usual gradient of the shape functions.
+                    const Tensor<2, dim, double> projection = unit_symmetric_tensor<dim>() -
+                                                              outer_product(fs_fe_face_values.normal_vector(point), fs_fe_face_values.normal_vector(point));
+
+                    // Loop over the shape functions
+                    for (unsigned int i=0; i<dofs_per_cell; ++i)
+                      {
+                        // Make sure we only assemble for the y-component
+                        // TODO this is only correct for box geometries
+                        if (mesh_deformation_dof_handler.get_fe().system_to_component_index(i).first != dim-1)
+                          continue;
```

Comment:
same here -- add an assertion of some sort

Reply:
There is an AssertThrow for Box geometries in the `initialize` function, isn't that enough?

Reply:
Once someone goes around and fixes the first place, these secondary places still have to be fixed, but one might not notice if the first assertion is removed.

I see assertions as basically a code expression of what I assume to be true for each piece of code. If you make the same assumption in 7 different functions, then there should be 7 assertions that check that the assumption is true. It will make fixing this at a later time so much easier, because you will have already marked up all of the places where someone will have to go check.